### DISCUSSION TITLE
Improve uplink_br resilence

### DISCUSSION THREAD

**Discussion by amarpad** [magma/magma#7991] [Ideas]
During pipelined restart, self._delete_all_flows() is called on the uplink_br. This is super risky as everytime pipelined restarts we have a chance of bricking the box if flows don't restore properly.
Due to an upgrade issue from 1.3.3 on pipelined.yml not getting clobbered (I think it is because there was a local change to pipelined.yml on the box), this caused pipelined to not come up (some mismatch in enforcement config) and black hole the box as the interface didn't come up.

I think as a hot patch we should update uplink_bridge.py to check if the flows have changed and only then affect the uplink otherwise this is too risky, plus it kinda breaks stateless as we are affecting datapath traffic while we restart pipelined.

**Comment by amarpad** [comment]
cc @pshelar @koolzz @talkhasib @ulaskozat @jpad-freedomfi

**Comment by pshelar** [comment]
posted PR: https://github.com/magma/magma/pull/7997

**Comment by pshelar** [comment]
did you saw following flow on the uplink_br0?
```
 cookie=0x0, duration=61.409s, table=0, n_packets=23, n_bytes=1862, priority=0 actions=NORMAL
```

**Reply by pshelar** [reply]
I could reproduce the issue, here is what I see:
1. after restarting 1.5.1 from 1.3.3 config pipelineD is crashing due to following exception:
```
Jul 13 02:57:12 phy-u17 pipelined[731644]: Traceback (most recent call last):
Jul 13 02:57:12 phy-u17 pipelined[731644]:   File "/usr/lib/python3.8/runpy.py", line 194, in _run_module_as_main
Jul 13 02:57:12 phy-u17 pipelined[731644]:     return _run_code(code, main_globals, None,
Jul 13 02:57:12 phy-u17 pipelined[731644]:   File "/usr/lib/python3.8/runpy.py", line 87, in _run_code
Jul 13 02:57:12 phy-u17 pipelined[731644]:     exec(code, run_globals)
Jul 13 02:57:12 phy-u17 pipelined[731644]:   File "/usr/local/lib/python3.8/dist-packages/magma/pipelined/main.py", line 190, in <module>
Jul 13 02:57:12 phy-u17 pipelined[731644]:     main()
Jul 13 02:57:12 phy-u17 pipelined[731644]:   File "/usr/local/lib/python3.8/dist-packages/magma/pipelined/main.py", line 110, in main
Jul 13 02:57:12 phy-u17 pipelined[731644]:     service_manager.load()
Jul 13 02:57:12 phy-u17 pipelined[731644]:   File "/usr/local/lib/python3.8/dist-packages/magma/pipelined/service_manager.py", line 573, in load
Jul 13 02:57:12 phy-u17 pipelined[731644]:     for app in manager.instantiate_apps(**contexts):
Jul 13 02:57:12 phy-u17 pipelined[731644]:   File "/usr/lib/python3/dist-packages/ryu/base/app_manager.py", line 509, in instantiate_apps
Jul 13 02:57:12 phy-u17 pipelined[731644]:     self._instantiate(app_name, cls, *args, **kwargs)
Jul 13 02:57:12 phy-u17 pipelined[731644]:   File "/usr/lib/python3/dist-packages/ryu/base/app_manager.py", line 495, in _instantiate
Jul 13 02:57:12 phy-u17 pipelined[731644]:     app = cls(*args, **kwargs)
Jul 13 02:57:12 phy-u17 pipelined[731644]:   File "/usr/local/lib/python3.8/dist-packages/magma/pipelined/app/enforcement_stats.py", line 99, in __init__
Jul 13 02:57:12 phy-u17 pipelined[731644]:     kwargs['config']['enforcement']['default_drop_flow_name']
Jul 13 02:57:12 phy-u17 pipelined[731644]: KeyError: 'default_drop_flow_name'
```
2. This crash loop is detected by agw service monitoring and it does redis flush after couple of minutes of crash loop
3. The redis flush also cleanup the OVS state which results in empty flow table.

I will post PR that save and restores uplink_br0 flow table so that AGW is still accessible over SGi interface.

**Reply by amarpad** [reply]
NICE!!! awesome if you can cherry pick it to 1.5 as we only upgraded one site.

**Reply by pshelar** [reply]
PR: https://github.com/magma/magma/pull/8146
